Update UG with redo and undo features (#154)
AY2223S1-CS2103T-T10-3/tp: docs/UserGuide.md

### 6.7. Undoing your previous command:

What if you accidentally made a mistake and performed a wrong <em>undoable</em> command?
Workbook can help you to undo your changes after performing an undesirable <em>undoable</em> command! 
Workbook will restore itself to the version before you performed the previous <em>undoable</em> command!

> WorkBook adds a new version for an <em>undoable</em> command and keeps track of all its previous versions, 
> allowing the `undo` command to be performed as many times to restore the WorkBook to any of its previous tracked versions. 

Reviewer: ```suggestion
WorkBook keeps track of all your previous commands, allowing you to `undo` as many times to restore any desired state of your WorkBook. 
```

version of the WorkBook.
* `edit 2 l/golang t/` followed by `delete 1` and then 2 consecutive `undo` commands will first restore the
WorkBook to its version before the `delete 1` command, and then the version before the `edit` command.

Before executing command: `edit 2 l/golang t/`
![BeforeEdit](images/BeforeEdit.png)

After executing command: `edit 2 l/golang t/`
![AfterEdit](images/AfterEdit.png)

After executing command: `delete 1`
![AfterDelete](images/AfterDelete.png)

After executing command: `undo`
![AfterUndo1](images/AfterUndo1.png)

After executing command: `undo`
![AfterUndo2](images/AfterUndo2.png)

<div markdown="block" class="alert alert-info">

**:information_source: Note:**<br>
You can only undo <em>undoable</em> commands!
* `add`
* `edit`
* `delete`
* `clear`
* `redo`
</div>

<div markdown="block" class="alert alert-success">

**:bulb: Tip:** <br>
* If you have undone your previous undoable command by mistake, fret not as you can easily redo this 
to reapply the initial changes!
* Performing multiple `undo` commands may confuse you as to which <em>undoable</em> command you are

> undoing and how your WorkBook currently looks like!

Reviewer: i think dunnid include this as it may look like users are expecting that they can change states manually, ygwim right, what do u think?

* Tracked WorkBook versions will reset once you `exit` the application!

</div>

### 6.8. Redoing your previous command:

Now, what if you mistakenly performed an `undo` command?
Workbook can help you to redo your changes after performing an undesirable `undo` command!
Workbook will reverse the previous `undo` command and restore itself to the version before that!

WorkBook keeps track of all your previous <em>undoable</em> commands and its current version,
allowing the `redo` command to be performed as many times to reverse the
`undo` commands you have performed!

Format: `redo`
* Redo the previous `undo` command.
* You can perform the redo command as many times as the number of undone versions in the Workbook's tracked versions
i.e. stacking 2 redo commands sequentially will revert the WorkBook back to its version 2
  `undo` commands ago, assuming that you performed at least 2 `undo` commands previously.
* When you perform an `undo` command, it will be stacked on top of the 
previous `undo` command in the stack. The `redo` command will redo the first `undo` command at the
top of the stack.

> * If you have not executed a previous `undo` command, the WorkBook will remain in its current state

Reviewer: lets explain the definition of "state" in the "Glossary" section

### 6.7. Undoing your previous command:

If you executed a wrong <em>undoable</em> command or changed your mind after executing an <em>undoable</em> command, this command 

> will restore the WorkBook to its earlier state before the previous <em>undoable</em> command was executed.

Reviewer: How to tell if my command is undo-able? maybe an example to elaborate will be useful!

And also kind of tough to digest, maybe something simpler like:
I accidentally_ cleared my list of internship applications, what do I do??
WorkBook helps undo/revert your changes if you happen to change your mind after performing an undesirable action ......

WorkBook adds a new state for an <em>undoable</em> command and uses a list to keep track of all its previous states, allowing the `undo` command to be
executed as many times to restore the WorkBook to any of its previous tracked state. 

Format: `undo`
* Undo the previous <em>undoable</em> command.
* You can execute the `undo` command as many times as the number of states in the Workbook's list 
of tracked states i.e. stacking 2 undo commands sequentially will revert the WorkBook back to its state 2
states ago, assuming the WorkBook has executed at least 2 <em>undoable</em> commands previously.
* If you have not executed an <em>undoable</em> command previously, the WorkBook will remain in its current state 
and return an error message: "No previous changes to undo!".
* <em>Undoable</em> commands are stacked, when an <em>undoable</em> command is executed, it will be stacked on top of the
previous <em>undoable</em> command in the stack. The `undo` command will undo the first <em>undoable</em> command at the

> top of the stack.

Reviewer: Some images will be nice :)

Examples:
* `add c/Meta r/Frontend Engineer s/Application Sent d/29-Oct-2022 12:00 e/hrmonkey@example.com` followed by `undo` will
return the WorkBook state without the added internship.
* `edit 1 s/Behavioural Interview e/hr@meta.com` followed by `clear` and then `undo` will undo the 
clearing of the Workbook. The edited stage and email address of the first internship will remain in the 
returned WorkBook state.
* `edit 2 l/golang t/` followed by `delete 1` and then 2 consecutive `undo` commands will first restore the 
WorkBook to its state before the `delete 1` command was executed, and then to its state before the `edit` command 
was executed.
* `delete 2` followed by `list` and then `undo` will restore the WorkBook to its state prior to the 
`delete 2` command as the `list` command is not an <em>undoable</em> command and does not change the 

> state of the WorkBook.

Reviewer: similar to comment above, outcome pics

<div markdown="block" class="alert alert-info">

**:information_source: Note:**<br>
You can only undo <em>undoable</em> commands!
* `add`
* `edit`
* `delete`
* `clear`
* `redo`

> </div>

Reviewer: ohh so its here

<div markdown="block" class="alert alert-success">

**:bulb: Tip:** <br>
* If you have undone your previous undoable command by mistake, fret not as you can easily redo this 
to reapply the initial changes!
* Stacking multiple `undo` commands may confuse you as to which <em>undoable</em> command you are

> undoing and which state your WorkBook is in!

Reviewer: let's keep the user in mind again, some may not know what 'state' is (especially those control F ppl), so u can refer to the comment i made above about `undo` and try tweaking this to make it more user-friendly

* `undo` commands are stacked, when an `undo` command is executed, it will be stacked on top of the 
previous `undo` command in the stack. The `redo` command will redo the first `undo` command at the
top of the stack.

Examples:
* `add c/Meta r/Frontend Engineer s/Application Sent d/29-Oct-2022 12:00 e/hrmonkey@example.com` followed by `undo` 
and then `redo` will restore the WorkBook to its state with the added internship.
* `edit 1 s/Behavioural Interview e/hr@meta.com` followed by `clear` will create 2 new states 
to the WorkBook. Following 2 consecutive `undo` commands, the WorkBook will be reverted back to 
its original state before both `edit` and `clear` commands. When a `redo` command is executed now,
the WorkBook will return the state with the edited stage and email for the first internship.

<div markdown="block" class="alert alert-success">

**:bulb: Tip:** <br>
* If you have redone your previous undo command by mistake, fret not as you can easily undo this again
to remove the initial changes!
* Stacking multiple `redo` commands may confuse you as to which <em>undoable</em> command you are
undoing and which state your WorkBook is in!

> </div>

Reviewer: same, some images wld be useful to illustrate ur points
